utf8/reader_test.go

```diff
@@ -0,0 +1,288 @@
+			repeat: 1024,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 1025,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2048,
+		},
+		{
+			input:  "Εν οίδα ότι ουδέν οίδα",
+			repeat: 2049,
+		},
+		{
+			input:  string([]byte{206, 149, 206, 206, 206}),
+			repeat: 1,
+			err:    utf8.ErrDecode,
+			errOut: []rune{'Ε'},
+		},
+	}
+
+	// test NewReaderReader
+	for _, tc := range testcases {
+		reader := utf8.NewReaderReader(iotest.NewRepeatReader(
+			bytes.NewBuffer([]byte(tc.input)),
+			tc.repeat))
+
+		got, readErr := runes.ReadAll(reader)
+		assert.IsError(t, tc.err, readErr, "Read() error")
+
+		var expected []rune
+
+		if readErr == nil {
+			repeater := iotest.NewRepeatReader(bytes.NewBuffer([]byte(tc.input)),
+				tc.repeat)
+
+			expectedBytes, err := io.ReadAll(repeater)
+			assert.NoError(t, err, "repeating expected")
+
+			expected = []rune(string(expectedBytes))
+		} else {
+			expected = tc.errOut
+		}
+
+		assert.EqualInts(t, len(expected), len(got), "rune slice len mismatch: %s",
+			string(got))
+
+		for i, r := range expected {
+			if r != got[i] {
+				t.Errorf("want[%d = %c] but got[%d = %c]", r, r, got[i], got[i])
+			}
+		}
+	}
+
+	// test NewReader
+	for _, tc := range testcases {
+		inputReader := iotest.NewRepeatReader(
+			bytes.NewBuffer([]byte(tc.input)),
+			tc.repeat)
+
+		data, err := io.ReadAll(inputReader)
+		assert.NoError(t, err, "reading repeated string")
+
+		reader := utf8.NewReader(bytes.NewBuffer(data))
+
+		got, readErr := runes.ReadAll(reader)
+		assert.IsError(t, tc.err, readErr, "read() error")
+
+		var expected []rune
+
+		if readErr == nil {
+			repeater := iotest.NewRepeatReader(bytes.NewBuffer([]byte(tc.input)),
+				tc.repeat)
+
+			expectedBytes, err := io.ReadAll(repeater)
+			assert.NoError(t, err, "repeating expected")
+
+			expected = []rune(string(expectedBytes))
+		} else {
+			expected = tc.errOut
+		}
+
+		assert.EqualInts(t, len(expected), len(got), "rune slice len mismatch")
+
+		for i, r := range expected {
+			if r != got[i] {
+				t.Errorf("want[%d = %c] but got[%d = %c]", r, r, got[i], got[i])
+			}
+		}
+	}
+
+	// TODO(i4k): implement stdiotest.TestReader for utf8.Reader
+}
+
+func TestUTF8ReaderMultipleSizes(t *testing.T) {
+	input := iotest.NewRepeatReader(bytes.NewBuffer([]byte(expectedString)), 100)
+	inputBytes, err := io.ReadAll(input)
+	assert.NoError(t, err, "failed to repeat input")
+
+	for i := 0; i < 10; i++ {
+		buf1 := bytes.NewBuffer(inputBytes)
+		buf2 := bytes.NewBuffer(inputBytes)
+		repeater1 := iotest.NewRepeatReader(buf1, i)
+		repeater2 := iotest.NewRepeatReader(buf2, i)
+
+		runes, err := runes.ReadAll(utf8.NewReaderReader(repeater1))
+		assert.NoError(t, err, "reading runes")
+
+		expectedBytes, err := io.ReadAll(repeater2)
+		assert.NoError(t, err, "reading expected runes")
+
+		expected := []rune(string(expectedBytes))
+
+		assert.EqualInts(t, len(expected), len(runes), "length mismatch")
+
+		for j := 0; j < len(expected); j++ {
```

Comment:
hmm since this is the third time I see a rune slice assertEqual thing going on I think we could extract this on a local assertEqualRunes or something like it.